File: USACO/Silver/Python/USACO 2016 Jan Silver Subsequences Summing to Seven.py
import sys

sys.stdin = open('div7.in', 'r')
sys.stdout = open('div7.out', 'w')

input = sys.stdin.readline

n = int(input())
cows = [int(input()) for _ in range(n)]

pref = [0] * (n+1)

for i in range(n):
    pref[i+1] = pref[i] + cows[i]%7
    pref[i+1] %= 7

ans = 0
earliest = {}
for i in range(1, n+1):
    if pref[i] in earliest:
        ans = max(ans, i - earliest[pref[i]])
    else:
        earliest[pref[i]] = i

#print(cows)
#print(pref)
print(ans)

File: USACO/Silver/C++/USACO 2016 Jan Silver Subsequences Summing to Seven.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	int n;
	cin >> n;
	int pref[n+1];
	pref[0] = 0;
	for (int i=0; i<n; ++i){
		int id; cin >> id;
		pref[i+1] = (pref[i] + id) % 7;
	}

	int firstOcc[7];
	fill(firstOcc, firstOcc+7, -1);

	long long ans = 0;
	for (int i=0; i<n+1; ++i){
		if (firstOcc[pref[i]] == -1){
			firstOcc[pref[i]] = i;
		}
		else{
			ans = max(ans, (long long)i - firstOcc[pref[i]]);
		}
	}
	cout << ans;
}
